queries/ja.py

```diff
@@ -0,0 +1,297 @@
+"""
+
+    Greynir: Natural language processing for Icelandic
+
+    Copyright (C) 2021 Miðeind ehf.
+
+       This program is free software: you can redistribute it and/or modify
+       it under the terms of the GNU General Public License as published by
+       the Free Software Foundation, either version 3 of the License, or
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module handles Ja.is API queries.
+
+"""
+
+# TODO: Handle context queries
+# TODO: Handle "hver er *heima*síminn hjá X"
+# TODO: Smarter disambiguation interaction
+
+from typing import Dict, Optional
+
+import re
+import logging
+from urllib.parse import urlencode
+from datetime import datetime, timedelta
+
+from reynir import NounPhrase
+
+from . import query_json_api, gen_answer, numbers_to_neutral
+from query import Query
+from geo import iceprep_for_street
+from util import read_api_key
+
+
+# Module priority
+PRIORITY = 5
+
+# This module wants to handle parse trees for queries
+HANDLE_TREE = True
+
+# The grammar nonterminals this module wants to handle
+QUERY_NONTERMINALS = {"QJaQuery"}
+
+# The context-free grammar for the queries recognized by this plug-in module
+GRAMMAR = """
+
+Query →
+    QJaQuery '?'?
+
+QJaQuery →
+    QJaPhoneNumQuery
+
+QJaPhoneNumQuery →
+    QJaName4PhoneNumQuery | QJaPhoneNum4NameQuery
+
+QJaPhoneNum4NameQuery →
+    QJaWhatWhich "er" QJaTheNumber_nf "hjá" QJaSubject
+    | "hvaða" QJaTheNumber_þf "er" QJaSubject "með"
+    | "flettu" "upp" QJaTheNumber_þgf "hjá" QJaSubject
+
+QJaName4PhoneNumQuery →
+    "hver" "er" "með" QJaTheNumber_þf QJaPhoneNum QJaInPhonebook?
+    | "hverjir" "eru" "með" QJaTheNumber_þf QJaPhoneNum QJaInPhonebook?
+    | "flettu" "upp" QJaTheNumber_þgf QJaPhoneNum QJaInPhonebook?
+    | "flettu" "upp" QJaPhoneNum QJaInPhonebook
+    | "hver" "er" "í" "síma" QJaPhoneNum
+
+QJaThemÞgf →
+    "hann" | "hana" | "þau" | "þá" | "þær" | "það" "númer"? | "þetta" "númer"?
+
+QJaInPhonebook →
+    "í" "símaskránni" | "á" "já" "punktur" "is"
+
+QJaPhoneNum →
+    Nl
+
+QJaSubject →
+    Nl
+
+QJaTheNumber/fall →
+    'númer:hk'/fall
+    | 'símanúmer:hk'/fall
+    | 'sími:kk'/fall
+
+QJaWhatWhich →
+    "hvert" | "hvað" | "hver"
+
+$score(+35) QJaQuery
+
+"""
+
+
+def QJaSubject(node, params, result):
+    n = result._text
+    nom = NounPhrase(n).nominative or n
+    result.qkey = nom
+
+
+def QJaPhoneNum(node, params, result):
+    result.phone_number = result._text
```

Comment:
Might be better to cast to nominative with result._nominative instead of result._text